docs/src/developing/runtime-facilities/zk-token-proof.md

```diff
@@ -0,0 +1,123 @@
+ciphertext can then be decrypted using a corresponding ElGamal secret key. The
+variant that is used in the proof program is the
+[twisted ElGamal encryption](https://eprint.iacr.org/2019/319) over the elliptic
+curve [curve25519](https://www.rfc-editor.org/rfc/rfc7748#section-4.1).
+
+The Pedersen commitment scheme is a popular instantiation of a cryptographic
+commitment scheme. A commitment scheme allows a user to wrap a message into a
+commitment with a purpose of revealing the committed message later on. Like a
+ciphertext, the resulting commitment does not reveal any information about the
+containing message. At the same time, the commitment is binding in that the user
+cannot change the original value that is contained in a commitment.
+
+Interested readers can refer to the following resources for a more in-depth
+treatment of Pedersen commitment and the (twisted) ElGamal encryption schemes.
+
+- [Notes](./zk-docs/twisted_elgamal.pdf) on the twisted ElGamal encryption
+- A technical
+  [overview](https://github.com/solana-labs/solana-program-library/blob/master/token/zk-token-protocol-paper/part1.pdf)
+  of the SPL Token 2022 confidential extension
+- Pretty Good Confidentiality [research paper](https://eprint.iacr.org/2019/319)
+
+The ZK Token proof program contains proof verification instructions on various
+zero-knowledge proofs for working with the Pedersen commitment and ElGamal
+encryption schemes. For example, the `VerifyRangeProofU64` instruction verifies
+a zero-knowledge proof certifying that a Pedersen commitment contains an
+unsigned 64-bit number as the message. The `VerifyPubkeyValidity` instruction
+verifies a zero-knowledge proof certifying that an ElGamal public key is a
+properly formed public key.
+
+### Context Data
+
+The proof data associated with each of the ZK Token proof instructions are
+logically divided into two parts:
+
+- The <em>context</em> component contains the data that a zero-knowledge proof
+  is certifying. For example, context component for a `VerifyRangeProofU64`
+  instruction data is the Pedersen commitment that holds an unsigned 64-bit
+  number. The context component for a `VerifyPubkeyValidity` instruction data is
+  the ElGamal public key that is properly formed.
+- The <em>proof</em> component contains the actual mathematical pieces that
+  certifies different properties of the context data.
+
+The ZK Token proof program processes a proof instruction in two steps:
+
+1. Verify the zero-knowledge proof data associated with the proof instruction.
+2. If specified in the instruction, the program stores the context data in a
+   dedicated context state account.
+
+The simplest way to use a proof instruction is to execute it without producing a
+context state account. In this case, the proof instruction can be included as
+part of a larger Solana transaction that contains instructions of other Solana
+programs. Programs such as the SPL Token 2022 program should directly access the
```

Comment:
Do we need to call out SPL Token 2022 here?

Reply:
Hm, yeah I guess it is not necessary.

In general, I mention Token 2022 in the doc because:
1. It explains the specific choice of instructions in the proof program
2. People can perhaps look at Token 2022 to see an example of how proof program can be used

Here, I was intending the effect of (2), but perhaps we can remove the Token 2022 in this sentence and then we can mention to refer to Token 2022 for examples on how proof program can be used somewhere in the doc.

Reply:
Okay, I removed mentioning of Token 2022 here. I had a couple of applications like these https://github.com/solana-labs/solana-program-library/issues/3984 in mind that would need to use the proof program. Perhaps we can update doc with pointers to examples that include Token-2022 and some of these applications once they are complete.